flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResourceManagerPartitionTrackerImplTest.java

```diff
@@ -0,0 +1,241 @@
+
+		tracker.processTaskExecutorClusterPartitionReport(TASK_EXECUTOR_ID_1, EMPTY_PARTITION_REPORT);
+		assertThat(partitionReleaser.releaseCalls, empty());
+	}
+
+	/**
+	 * Verifies that a task executor hosting multiple partitions of a data set receives a release call if a subset of
+	 * its partitions is lost.
+	 */
+	@Test
+	public void testReportProcessingWithPartitionLossOnSameTaskExecutor() {
+		TestClusterPartitionReleaser partitionReleaser = new TestClusterPartitionReleaser();
+		final ResourceManagerPartitionTracker tracker = new ResourceManagerPartitionTrackerImpl(partitionReleaser);
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_1,
+			createClusterPartitionReport(DATA_SET_ID, 2, PARTITION_ID_1, PARTITION_ID_2));
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_1,
+			createClusterPartitionReport(DATA_SET_ID, 2, PARTITION_ID_2));
+
+		assertThat(partitionReleaser.releaseCalls, contains(Tuple2.of(TASK_EXECUTOR_ID_1, Collections.singleton(DATA_SET_ID))));
+	}
+
+	/**
+	 * Verifies that a task executor hosting partitions of a data set receives a release call if a partition of the
+	 * data set is lost on another task executor.
+	 */
+	@Test
+	public void testReportProcessingWithPartitionLossOnOtherTaskExecutor() {
+		TestClusterPartitionReleaser partitionReleaser = new TestClusterPartitionReleaser();
+		final ResourceManagerPartitionTracker tracker = new ResourceManagerPartitionTrackerImpl(partitionReleaser);
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_1,
+			createClusterPartitionReport(DATA_SET_ID, 2, PARTITION_ID_1));
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_2,
+			createClusterPartitionReport(DATA_SET_ID, 2, PARTITION_ID_2));
+
+		tracker.processTaskExecutorClusterPartitionReport(TASK_EXECUTOR_ID_1, EMPTY_PARTITION_REPORT);
+
+		assertThat(partitionReleaser.releaseCalls, contains(Tuple2.of(TASK_EXECUTOR_ID_2, Collections.singleton(DATA_SET_ID))));
+	}
+
+	@Test
+	public void testListDataSetsBasics() {
+		TestClusterPartitionReleaser partitionReleaser = new TestClusterPartitionReleaser();
+		final ResourceManagerPartitionTracker tracker = new ResourceManagerPartitionTrackerImpl(partitionReleaser);
+
+		assertEquals(0, tracker.listDataSets().size());
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_1,
+			createClusterPartitionReport(DATA_SET_ID, 1, PARTITION_ID_1));
+
+		final Map<IntermediateDataSetID, DataSetMetaInfo> listing = tracker.listDataSets();
+		assertThat(listing, hasKey(DATA_SET_ID));
+		DataSetMetaInfo metaInfo = listing.get(DATA_SET_ID);
+		assertEquals(1, metaInfo.getNumTotalPartitions());
+
+		tracker.processTaskExecutorClusterPartitionReport(
+			TASK_EXECUTOR_ID_1,
+			EMPTY_PARTITION_REPORT);
+		assertEquals(0, tracker.listDataSets().size());
+	}
+
+	@Test
+	public void testListDataSetsMultiplePartitionsOnSingleTaskExecutor() {
```

Comment:
minor: this test tests 3 transitions:
incomplete -> incomplete
incomplete -> complete
complete -> incomplete
I would prefer 3 test cases

some other tests also test multiple things

Reply:
Just an idea, if we init `partitionReleaser/tracker` in test setup and factor out
`tracker.processTaskExecutorClusterPartitionReport(TASK_EXECUTOR_ID_1, createClusterPartitionReport(..`
into e.g. `report(TASK_EXECUTOR_ID_1, DATA_SET_ID, 2, PARTITION_ID_1, ..)`
this would reduce code size and simplify writing tests with duplicated setup but testing one thing.

Reply:
I've addressed your second comment (I think).

As for the other one, your state transitions aren't quite correct, or rather overly simplified.
It goes from `unknown -> partially-complete -> complete -> partially-complete`.

Trying to split these up just means we duplicate code we already have. You can't go `partially-complete -> complete` without first having to go `unknown -> partially-complete`, so if you were to write separate tests you're running the same code `unknown -> partially-complete` twice.
And this is true for all state transitions.
Assertions are usable enough for us to differentiate between failures at different stages, so I don't buy in to the whole "test exactly one thing" mentality.

Reply:
Indeed, there is going to be some code duplication. The idea was to reduce its size with the `report` method (which happened) in favour of isolated checks at the end. I am ok to keep it like this.